File: js/merge_sort.js
function mergeSort(arr){
    if(arr.length <= 1) return arr;

    const mid = Math.floor(arr.length / 2);
    console.log(mid);
    
    const left = mergeSort(arr.slice(0, mid)); 
    const right = mergeSort(arr.slice(mid));
    // console.log(arr[mid]);
    
    console.log(left);
    console.log(right);
    
    return merge(left, right);
}

function merge(left, right){
    const merged = [];
    let i = j = 0;

    //merged two sorted values
    while (i < left.length && j<right.length){
        if(left[i] <= right[j]){
            merged.push(left[i]);
            console.log("i", left[i]);
            
            i++;
        }else{
            merged.push(right[j]);
            console.log("j", right[j]);
            j++;
        }
    }

    //add remaining elements
    return merged.concat(left.slice(i)).concat(right.slice(j));

}



const arr = [38, 27, 43, 3, 9, 82, 10, 23];
const sorted = mergeSort(arr);
console.log(sorted);

File: merge_sort.py
def mergeSort(arr):
    if len(arr) <= 1:
        return arr
    
    mid = len(arr) // 2
    left = mergeSort(arr[:mid])
    print(arr[:mid])
    right = mergeSort(arr[mid:])
    print(arr[mid:])

    return merge(left, right)


def merge(left, right):
    merged = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1
    
    merged.extend(left[i:])
    merged.extend(right[j:])

    return merged


arr = [38, 27, 43, 3, 9, 82, 10]
sorted_arr = mergeSort(arr)
print("Sorted:", sorted_arr)
